src/main/java/org/broadinstitute/barclay/argparser/CommandLinePluginDescriptor.java

```diff
@@ -76,103 +110,127 @@
     public abstract List<String> getPackageNames();
 
     /**
-     * Give this descriptor a chance to filter out any classes it doesn't want to be
-     * dynamically discoverable.
-     * @return false if the class shouldn't be used; otherwise true
+     * Base class for all command line plugin classes managed by this descriptor. Subclasses of
+     * this class in any of the packages returned by {@link #getPackageNames} will be command line
+     * accessible.
      */
-    public Predicate<Class<?>> getClassFilter() { return c -> true;}
+    public abstract Class<?> getPluginBaseClass();
 
     /**
-     * Return an instance of the specified pluggable class. The descriptor should
-     * instantiate or otherwise obtain (possibly by having been provided an instance
-     * through the descriptor's constructor) an instance of this plugin class.
-     * The descriptor should maintain a list of these instances so they can later
-     * be retrieved by {@link #getAllInstances}.
+     * Determine if a plugin class should be retained for this descriptor. Return true if the
+     * descriptor wants this class to be included in from the list of plugins discovered dynamically.
      *
-     * In addition, implementations should recognize and reject any attempt to instantiate
-     * a second instance of a plugin that has the same simple class name as another plugin
-     * controlled by this descriptor (which can happen if they have different qualified names
-     * within the base package used by the descriptor) since the user has no way to disambiguate
-     * these on the command line).
+     * @return false if the plugin class shouldn't be used; otherwise true
+     */
+    public boolean includePluginClass(Class<?> c) { return true;}
```

Comment:
I think that this is wrong, because to include a class for a plugin it should be at least of class `T`, which may be reached by bounding the parameter or check if it is assignable from the base class... Otherwise, every class might be included, even if it is not of class `T`.

Reply:
We can't use a type bound here, even a lower bound, since it would preclude the Barclay CLP or any code that doesn't have compile time knowledge of T from calling the method. We COULD use isAssignableFrom in the default implementation as a substitute, if thats what you're suggesting. It would never fail though, unless there was a bug in Barclay, which it would mask. It does seem more correct though.

Reply:
Changed default implementation to use isAssignableFrom.